docs/proposals/control-data-plane-split/README.md

```diff
@@ -0,0 +1,263 @@
+- If no nginx instances exist, control plane should not send any configuration.
+- The control plane should check if a connection exists first before sending the config.
+- If the control plane is scaled, then we should mark non-leaders as Unready (return non-200 readiness probe). This will prevent nginx agents from connecting to the non-leaders (k8s removes the Unready Pods from the Endpoint pool), and therefore only the leader will send config and write statuses.
+  - We will need to ensure that the leader Pod can handle many nginx connections.
+
+<img src="graph-conns.png" alt="Scaled connections" width=500 style="display: block; margin: 0 auto">
+
+### Agent API
+
+The control plane will need to implement the proper gRPC APIs in order to connect and send configuration to the agent.
+
+Useful references:
+
+- [Basic mock control plane](https://github.com/nginx/agent/tree/v3/test/mock/grpc)
+- [Proto definitions](https://github.com/nginx/agent/blob/v3/docs/proto/protos.md)
+
+The gRPC services to implement in the control plane are:
+
+`CommandService`: this handles the connection and subscription to the agent. This will make a `ConfigApplyRequest` when we need to write nginx config. This request will contain a `FileOverview`, which is essentially a list of filenames that are to be sent.
+
+`FileService`: when the agent receives a `ConfigApplyRequest`, it sends a `GetFile` request to this service to download the file contents contained in the original `FileOverview`.
+
+Some API methods will not need to be implemented. For example, `UpdateFile` is used by the agent to send a file that was updated on the agent side to the control plane. In our case, this won't happen since our control plane has full control over the config. Methods relating to this functionality can be stubbed out.
+
+### Metrics
+
+Agent can be configured to expose metrics on a `/metrics` endpoint. Our control plane will need to configure agent to do this so that Prometheus can scrape each agent for nginx metrics.
+
+### NGINX Plus
+
+#### Upstream server state
+
+In the current implementation using NGINX Plus, when only upstream servers change, NGF writes the upstream servers in the nginx config, but does not reload. It then calls the NGINX Plus API to update the servers. This process allows us to update the upstream servers using the API and not have to reload nginx, while still having the upstream servers exist in the nginx config for easy debugging and consistency. However, when using agent, any config write will result in a reload. To preserve the ability to update upstreams with the API without needing a reload, we'll have to utilize a `state` file instead of writing the servers directly in the nginx config. This way the list of servers is still available on the filesystem for debugging, but is written by the nginx when making the API call instead of by the control plane directly.
```

Comment:
Does the agent v3 API support this? If so, what's the API endpoint/payload look like?

Reply:
They are adding support for us to make the UpdateHTTPUpstreams API call. That's all we need, then NGINX Plus will handle the state file.

Not sure what the payload looks like yet, but I think it is or will be documented internally.

Reply:
Ok, let's add that context here.